File: tile_gen_proto/tool_palette_maker.cpp
//================================================================================================================================
//=> - Includes -
//================================================================================================================================

#include <cstdio>
#include <cstdlib>
#include <climits>
#include <vector>
#include <map>
#include <set>
#include <algorithm>
#include <random>
#include <string>
#include <fstream>
#include <queue>

//================================================================================================================================
//=> - Types and structures -
//================================================================================================================================

struct Color {
    unsigned char r, g, b;
    
    Color() : r(0), g(0), b(0) {}
    Color(unsigned char r_, unsigned char g_, unsigned char b_) : r(r_), g(g_), b(b_) {}
    
    bool operator<(const Color& other) const {
        if (r != other.r) return r < other.r;
        if (g != other.g) return g < other.g;
        return b < other.b;
    }
    
    bool operator==(const Color& other) const {
        return r == other.r && g == other.g && b == other.b;
    }
    
    bool operator!=(const Color& other) const {
        return !(*this == other);
    }
};

struct Point {
    int x, y;
    Point(int x_, int y_) : x(x_), y(y_) {}
};

//================================================================================================================================
//=> - Class: PaletteMaker -
//================================================================================================================================

class PaletteMaker {
private:
    std::string image_path;
    std::string save_path;
    std::string palette_img_save;
    int num_colors;
    int width;
    int height;
    std::vector<std::vector<Color>> image_array;
    std::map<Color, std::vector<Point>> color_to_coords;
    std::vector<Color> palette_colors;
    std::mt19937 rng;

    int colorDistance(const Color& color1, const Color& color2) {
        return abs((int)color1.r - (int)color2.r) + 
               abs((int)color1.g - (int)color2.g) + 
               abs((int)color1.b - (int)color2.b);
    }

    void loadImage() {
        FILE* file = fopen(image_path.c_str(), "rb");
        if (file == nullptr) {
            printf("*** Error: Could not open image file: %s\n", image_path.c_str());
            image_array.clear();
            return;
        }
        
        char magic[3];
        if (fread(magic, 1, 2, file) != 2 || magic[0] != 'P' || magic[1] != '6') {
            printf("*** Error: Not a valid PPM file (P6 format required): %s\n", image_path.c_str());
            fclose(file);
            image_array.clear();
            return;
        }
        magic[2] = '\0';
        
        int c;
        while ((c = fgetc(file)) != EOF && (c == ' ' || c == '\t' || c == '\n' || c == '\r')) {}
        if (c == '#') {
            while ((c = fgetc(file)) != EOF && c != '\n') {}
            c = fgetc(file);
        }
        ungetc(c, file);
        
        if (fscanf(file, "%d %d", &width, &height) != 2) {
            printf("*** Error: Could not read image dimensions from: %s\n", image_path.c_str());
            fclose(file);
            image_array.clear();
            return;
        }
        
        int max_val;
        if (fscanf(file, "%d", &max_val) != 1) {
            printf("*** Error: Could not read max value from: %s\n", image_path.c_str());
            fclose(file);
            image_array.clear();
            return;
        }
        
        fgetc(file);
        
        image_array.resize(height);
        for (int y = 0; y < height; y++) {
            image_array[y].resize(width);
            for (int x = 0; x < width; x++) {
                unsigned char r = fgetc(file);
                unsigned char g = fgetc(file);
                unsigned char b = fgetc(file);
                if (feof(file)) {
                    printf("*** Error: Unexpected end of file while reading pixel data\n");
                    fclose(file);
                    image_array.clear();
                    return;
                }
                image_array[y][x] = Color(r, g, b);
            }
        }
        
        fclose(file);
        analyzeColors();
    }

    void analyzeColors() {
        if (image_array.empty()) {
            return;
        }
        
        color_to_coords.clear();
        
        for (int y = 0; y < height; y++) {
            for (int x = 0; x < width; x++) {
                Color color = image_array[y][x];
                color_to_coords[color].push_back(Point(x, y));
            }
        }
        
        printf("*** Found %zu unique colors\n", color_to_coords.size());
        mergeColors();
    }

    void mergeColors() {
        if (image_array.empty() || color_to_coords.empty()) {
            return;
        }
        
        std::vector<Color> color_list;
        for (const auto& pair : color_to_coords) {
            color_list.push_back(pair.first);
        }
        
        int initial_count = (int)color_list.size();
        printf("*** Starting merge: %d colors -> %d colors\n", initial_count, num_colors);
        printf("*** Creating distance table for %d colors", initial_count);
        fflush(stdout);
        
        std::map<Color, std::vector<std::pair<int, Color>>> distance_lists;
        for (size_t i = 0; i < color_list.size(); i++) {
            Color color1 = color_list[i];
            distance_lists[color1] = std::vector<std::pair<int, Color>>();
            for (size_t j = 0; j < color_list.size(); j++) {
                if (i != j) {
                    Color color2 = color_list[j];
                    int distance = colorDistance(color1, color2);
                    distance_lists[color1].push_back(std::make_pair(distance, color2));
                }
            }
            std::sort(distance_lists[color1].begin(), distance_lists[color1].end());
        }
        
        printf("\n");
        
        int iteration = 0;
        while ((int)color_list.size() > num_colors) {
            if (distance_lists.empty()) {
                break;
            }
            
            int min_distance = INT_MAX;
            Color color1_min, color2_min;
            bool found = false;
            
            for (const auto& color_entry : distance_lists) {
                const Color& color = color_entry.first;
                const std::vector<std::pair<int, Color>>& dist_list = color_entry.second;
                if (!dist_list.empty()) {
                    int dist = dist_list[0].first;
                    if (dist < min_distance) {
                        min_distance = dist;
                        color1_min = color;
                        color2_min = dist_list[0].second;
                        found = true;
                    }
                }
            }
            
            if (!found) {
                break;
            }
            
            Color color1 = color1_min;
            Color color2 = color2_min;
            int count1 = (int)color_to_coords[color1].size();
            int count2 = (int)color_to_coords[color2].size();
            
            Color merge_from, merge_to;
            if (count1 < count2) {
                merge_from = color1;
                merge_to = color2;
            } else if (count2 < count1) {
                merge_from = color2;
                merge_to = color1;
            } else {
                std::uniform_real_distribution<double> dist(0.0, 1.0);
                if (dist(rng) < 0.5) {
                    merge_from = color1;
                    merge_to = color2;
                } else {
                    merge_from = color2;
                    merge_to = color1;
                }
            }
            
            distance_lists.erase(merge_from);
            
            for (auto& color_entry : distance_lists) {
                std::vector<std::pair<int, Color>>& dist_list = color_entry.second;
                dist_list.erase(
                    std::remove_if(dist_list.begin(), dist_list.end(),
                        [&merge_from](const std::pair<int, Color>& p) {
                            return p.second == merge_from;
                        }),
                    dist_list.end()
                );
            }
            
            color_to_coords[merge_to].insert(
                color_to_coords[merge_to].end(),
                color_to_coords[merge_from].begin(),
                color_to_coords[merge_from].end()
            );
            color_to_coords.erase(merge_from);
            
            color_list.erase(std::remove(color_list.begin(), color_list.end(), merge_from), color_list.end());
            
            iteration++;
            int remaining = (int)color_list.size();
            printf("*** Merging iteration %d: %d colors remaining\r", iteration, remaining);
            fflush(stdout);
        }
        
        printf("\n");
        
        std::vector<Color> final_colors;
        for (const auto& pair : color_to_coords) {
            final_colors.push_back(pair.first);
        }
        
        printf("*** Merge complete: %zu colors\n", final_colors.size());
        setPalette(final_colors);
        repaintImage();
        
        std::ofstream file(save_path);
        if (file.is_open()) {
            for (const auto& pair : color_to_coords) {
                Color color = pair.first;
                file << (int)color.r << ":" << (int)color.g << ":" << (int)color.b << "\n";
            }
            file.close();
        }
    }

    void repaintImage() {
        if (image_array.empty() || color_to_coords.empty()) {
            return;
        }
        
        std::vector<Color> palette_colors_list;
        for (const auto& pair : color_to_coords) {
            palette_colors_list.push_back(pair.first);
        }
        
        for (int y = 0; y < height; y++) {
            for (int x = 0; x < width; x++) {
                Color original_color = image_array[y][x];
                Color closest_color;
                int min_dist = INT_MAX;
                
                for (const Color& palette_color : palette_colors_list) {
                    int dist = colorDistance(original_color, palette_color);
                    if (dist < min_dist) {
                        min_dist = dist;
                        closest_color = palette_color;
                    }
                }
                
                image_array[y][x] = closest_color;
            }
        }
        
        saveRepaintedImage();
    }

    void saveRepaintedImage() {
        FILE* file = fopen(palette_img_save.c_str(), "wb");
        if (!file) {
            printf("*** Error: Could not open file for writing: %s\n", palette_img_save.c_str());
            return;
        }
        
        char header[64];
        int header_len = snprintf(header, sizeof(header), "P6\n%d %d\n255\n", width, height);
        fwrite(header, 1, header_len, file);
        
        for (int y = 0; y < height; y++) {
            for (int x = 0; x < width; x++) {
                Color color = image_array[y][x];
                fputc(color.r, file);
                fputc(color.g, file);
                fputc(color.b, file);
            }
        }
        
        fclose(file);
        printf("*** Saved repainted image to %s\n", palette_img_save.c_str());
    }

public:
    PaletteMaker(const std::string& image_path_, int num_colors_) 
        : image_path(image_path_), num_colors(num_colors_), rng(std::random_device{}()) {
        save_path = image_path;
        size_t pos = save_path.rfind(".png");
        if (pos != std::string::npos) {
            save_path.replace(pos, 4, "_palette.txt");
        } else {
            save_path += "_palette.txt";
        }
        
        palette_img_save = image_path;
        pos = palette_img_save.rfind(".png");
        if (pos != std::string::npos) {
            palette_img_save.replace(pos, 4, "_palette.ppm");
        } else {
            // Remove any existing extension and add .ppm
            size_t last_dot = palette_img_save.find_last_of(".");
            if (last_dot != std::string::npos) {
                palette_img_save = palette_img_save.substr(0, last_dot) + "_palette.ppm";
            } else {
                palette_img_save += "_palette.ppm";
            }
        }
        
        palette_colors.resize(num_colors, Color(0, 0, 0));
        loadImage();
    }

    void setPalette(const std::vector<Color>& colors) {
        palette_colors = colors;
        if ((int)palette_colors.size() < num_colors) {
            palette_colors.resize(num_colors, Color(0, 0, 0));
        } else if ((int)palette_colors.size() > num_colors) {
            palette_colors.resize(num_colors);
        }
    }

    std::vector<Color> getPalette() const {
        return palette_colors;
    }
};

//================================================================================================================================
//=> - Main -
//================================================================================================================================

int main(int argc, char* argv[]) {
    if (argc != 2) {
        printf("*** Error: Usage: %s <image_path>\n", argv[0]);
        return 1;
    }
    
    std::string image_path = argv[1];
    int num_colors = 8;
    PaletteMaker maker(image_path, num_colors);
    
    return 0;
}

//================================================================================================================================
//=> - End -
//================================================================================================================================

File: gen_tile_proto/tool_palette_maker.py
#================================================================================================================================#
#=> - Imports -
#================================================================================================================================#

import sys
sys.dont_write_bytecode = True

import os
os.chdir(os.path.dirname(os.path.abspath(__file__)))
sys.path.insert(0, os.getcwd())

import numpy as np
import random
import tkinter as tk
from PIL import Image, ImageTk

#================================================================================================================================#
#=> - Class: PaletteMaker
#================================================================================================================================#

class PaletteMaker:

    def __loadImage (m):
        if os.path.exists(m.image_path):
            img = Image.open(m.image_path)
            img_rgb = img.convert('RGB')
            m.image_array = np.array(img_rgb)
            m.__displayImage()
            m.__analyzeColors()
        else:
            m.image_array = None

    def __analyzeColors (m):
        if m.image_array is None:
            return
        unique_colors = set()
        color_to_coords = {}
        height, width = m.image_array.shape[:2]
        for y in range(height):
            for x in range(width):
                r, g, b = m.image_array[y, x]
                color_tuple = (int(r), int(g), int(b))
                unique_colors.add(color_tuple)
                if color_tuple not in color_to_coords:
                    color_to_coords[color_tuple] = []
                color_to_coords[color_tuple].append((x, y))
        m.color_to_coords = color_to_coords
        print ("*** Found %d unique colors" %(len(unique_colors)))
        m.__mergeColors()

    def __colorDistance (m, color1, color2):
        r1, g1, b1 = color1
        r2, g2, b2 = color2
        return abs(r1 - r2) + abs(g1 - g2) + abs(b1 - b2)

    def __mergeColors (m):
        if m.image_array is None or m.color_to_coords is None:
            return
        color_list = list(m.color_to_coords.keys())
        initial_count = len(color_list)
        print ("*** Starting merge: %d colors -> %d colors" %(initial_count, m.num_colors))
        print ("*** Creating distance table for %d colors" %(initial_count), end='\r')
        sys.stdout.flush()
        distance_table = {}
        for i in range(initial_count):
            for j in range(i + 1, initial_count):
                color1 = color_list[i]
                color2 = color_list[j]
                distance = m.__colorDistance(color1, color2)
                if color1 < color2:
                    distance_table[(color1, color2)] = distance
                else:
                    distance_table[(color2, color1)] = distance
        print ("")
        iteration = 0
        while len(color_list) > m.num_colors:
            if len(distance_table) == 0:
                break
            min_distance = min(distance_table.values())
            closest_pair = None
            for pair, dist in distance_table.items():
                if dist == min_distance:
                    closest_pair = pair
                    break
            if closest_pair is None:
                break
            color1, color2 = closest_pair
            count1 = len(m.color_to_coords[color1])
            count2 = len(m.color_to_coords[color2])
            if count1 < count2:
                merge_from, merge_to = color1, color2
            elif count2 < count1:
                merge_from, merge_to = color2, color1
            else:
                if random.random() < 0.5:
                    merge_from, merge_to = color1, color2
                else:
                    merge_from, merge_to = color2, color1
            keys_to_remove = []
            for pair in distance_table.keys():
                if merge_from in pair:
                    keys_to_remove.append(pair)
            for key in keys_to_remove:
                del distance_table[key]
            m.color_to_coords[merge_to].extend(m.color_to_coords[merge_from])
            del m.color_to_coords[merge_from]
            color_list.remove(merge_from)
            iteration += 1
            remaining = len(color_list)
            print ("*** Merging iteration %d: %d colors remaining" %(iteration, remaining), end='\r')
            sys.stdout.flush()
        print ("")
        final_colors = list(m.color_to_coords.keys())
        print ("*** Merge complete: %d colors" %(len(final_colors)))
        m.setPalette(final_colors)
        m.__repaintImage()
        with open(m.save_path, "w") as ptr:
            for color, coords in m.color_to_coords.items():
                ptr.write("%s\n" %(":".join([str(c) for c in color])))

    def __repaintImage (m):
        if m.image_array is None or m.color_to_coords is None:
            return
        height, width = m.image_array.shape[:2]
        repainted = np.zeros((height, width, 3), dtype=np.uint8)
        palette_colors = list(m.color_to_coords.keys())
        for y in range(height):
            for x in range(width):
                original_color = tuple(m.image_array[y, x])
                closest_color = None
                min_dist = float('inf')
                for palette_color in palette_colors:
                    dist = m.__colorDistance(original_color, palette_color)
                    if dist < min_dist:
                        min_dist = dist
                        closest_color = palette_color
                if closest_color is not None:
                    repainted[y, x] = closest_color
        m.image_array = repainted
        m.__displayImage()
        img = Image.fromarray(repainted, mode='RGB')
        img.save(m.palette_img_save)

    def __displayImage (m):
        m.image_canvas.delete("all")
        if m.image_array is not None:
            img = Image.fromarray(m.image_array, mode='RGB')
            photo = ImageTk.PhotoImage(img)
            canvas_width = m.image_canvas.winfo_width()
            canvas_height = m.image_canvas.winfo_height()
            if canvas_width <= 1:
                canvas_width = m.image_canvas.winfo_reqwidth()
            if canvas_height <= 1:
                canvas_height = m.image_canvas.winfo_reqheight()
            center_x = canvas_width // 2
            center_y = canvas_height // 2
            m.image_canvas.create_image(center_x, center_y, anchor=tk.CENTER, image=photo, tags="image")
            m.image_canvas.image = photo

    def __updatePaletteDisplay (m):
        for i, canvas in enumerate(m.palette_canvases):
            canvas.delete("all")
            if i < len(m.palette_colors):
                color = m.palette_colors[i]
                color_hex = "#%02x%02x%02x" % color
                canvas.config(bg=color_hex)

    def __init__ (m, root, image_path, num_colors):
        m.root = root
        m.image_path = image_path
        if image_path.endswith(".png"):
            m.save_path = image_path.replace(".png", "_palette.txt")
            m.palette_img_save = image_path.replace(".png", "_palette.png")
        elif image_path.endswith(".ppm"):
            m.save_path = image_path.replace(".ppm", "_palette.txt")
            m.palette_img_save = image_path.replace(".ppm", "_palette.ppm")
        else:
            raise ValueError("Unsupported image format: %s" %(image_path))
        m.save_path = image_path.replace(".png", "_palette.txt")
        m.palette_img_save = image_path.replace(".png", "_palette.png")
        m.num_colors = num_colors
        m.palette_colors = [(0, 0, 0)] * num_colors
        
        m.palette_frame = tk.Frame(root)
        m.palette_frame.pack(side=tk.TOP, fill=tk.X, padx=5, pady=5)
        
        m.palette_canvases = []
        color_canvas_size = 80
        for i in range(num_colors):
            canvas = tk.Canvas(m.palette_frame, width=color_canvas_size, height=color_canvas_size, bg="gray", highlightthickness=1, highlightbackground="black")
            canvas.pack(side=tk.LEFT, padx=2)
            m.palette_canvases.append(canvas)
        
        m.image_frame = tk.Frame(root)
        m.image_frame.pack(side=tk.TOP, fill=tk.BOTH, expand=True, padx=5, pady=5)
        
        m.image_canvas = tk.Canvas(m.image_frame, bg="#f0f0f0")
        m.image_canvas.pack(side=tk.TOP, fill=tk.BOTH, expand=True)
        m.image_canvas.bind("<Configure>", lambda e: m.__displayImage())
        
        m.__loadImage()
        m.__updatePaletteDisplay()

    def setPalette (m, colors):
        m.palette_colors = colors[:m.num_colors]
        if len(m.palette_colors) < m.num_colors:
            m.palette_colors.extend([(0, 0, 0)] * (m.num_colors - len(m.palette_colors)))
        m.__updatePaletteDisplay()

    def getPalette (m):
        return m.palette_colors.copy()

#================================================================================================================================#
#=> - Main -
#================================================================================================================================#

if __name__ == "__main__":
    if len(sys.argv) != 2:
        print ("*** Error: Usage: %s <image_path>" %(sys.argv[0]))
        sys.exit(1)
    image_path = sys.argv[1]
    
    root = tk.Tk()
    root.title("Palette Maker")
    num_colors = 8
    maker = PaletteMaker(root, image_path, num_colors)
    root.mainloop()

#================================================================================================================================#
#=> - End -
#================================================================================================================================#
